msal/src/test/java/com/microsoft/identity/client/e2e/tests/mocked/SingleAccountOverloadsMockedTest.java

```diff
@@ -0,0 +1,419 @@
+
+import static com.microsoft.identity.internal.testutils.TestConstants.Scopes.USER_READ_SCOPE;
+import static com.microsoft.identity.internal.testutils.mocks.MockTokenCreator.MOCK_PREFERRED_USERNAME_VALUE;
+import static org.junit.Assert.fail;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(shadows = {
+        ShadowStorageHelper.class,
+        ShadowAuthorityForMockHttpResponse.class,
+        ShadowHttpRequestForMockedTest.class,
+        ShadowMsalUtils.class,
+        ShadowOpenIdProviderConfigurationClient.class
+})
+public class SingleAccountOverloadsMockedTest extends AcquireTokenAbstractTest {
+
+    private SingleAccountPublicClientApplication mSingleAccountPCA;
+    private String mUsername = MOCK_PREFERRED_USERNAME_VALUE;
+
+    @Before
+    public void setup() {
+        super.setup();
+        TestUtils.clearCache(SingleAccountPublicClientApplication.SINGLE_ACCOUNT_CREDENTIAL_SHARED_PREFERENCES);
+        mSingleAccountPCA = (SingleAccountPublicClientApplication) mApplication;
+        MockHttpResponse.setHttpResponse(MockServerResponse.getMockTokenSuccessResponse());
+    }
+
+    @Test
+    public void testSignInOnlyAllowedOnce() {
+        mSingleAccountPCA.signIn(mActivity, mUsername, mScopes, getSuccessExpectedCallback());
+        RoboTestUtils.flushScheduler();
+
+        mSingleAccountPCA.signIn(mActivity, mUsername, mScopes, getInvalidParameterExpectedCallback());
+    }
+
+    @Test
+    public void testSignInWithPromptOnlyAllowedOnce() {
+        mSingleAccountPCA.signIn(mActivity, mUsername, mScopes, Prompt.LOGIN, getSuccessExpectedCallback());
+        RoboTestUtils.flushScheduler();
+
+        mSingleAccountPCA.signIn(mActivity, mUsername, mScopes, Prompt.LOGIN, getInvalidParameterExpectedCallback());
+    }
+
+    @Test
+    public void testSignInAgainAllowsSignInAgain() {
+        mSingleAccountPCA.signIn(mActivity, mUsername, mScopes, getSuccessExpectedCallback());
+        RoboTestUtils.flushScheduler();
+
+        mSingleAccountPCA.signInAgain(mActivity, mScopes, Prompt.LOGIN, getSuccessExpectedCallback());
+        RoboTestUtils.flushScheduler();
+    }
+
+    @Test
+    public void testCannotSignInAgainIfNeverSignedInBefore() {
+        mSingleAccountPCA.signInAgain(mActivity, mScopes, Prompt.LOGIN, getNoCurrentAccountExpectedCallback());
+    }
+
+    @Test
+    public void testCanSignOutIfAlreadySignedIn() {
+        mSingleAccountPCA.signIn(mActivity, mUsername, mScopes, getSuccessExpectedCallback());
+        RoboTestUtils.flushScheduler();
+
+        mSingleAccountPCA.signOut(new ISingleAccountPublicClientApplication.SignOutCallback() {
+            @Override
+            public void onSignOut() {
+                Assert.assertTrue("Successfully signed out", true);
+            }
+
+            @Override
+            public void onError(@NonNull MsalException exception) {
+                fail(exception.getMessage());
+            }
+        });
+
+        RoboTestUtils.flushScheduler();
```

Comment:
I see this at the conclusion of a few different tests; would this block make sense to put into an `@After` block?

(Also, I can't remember _why_ we need to flush the scheduler here; is this to reset thread state after test?)

Reply:
`flushScheduler` is being used in Robolectric tests to wait for the current runnable to finish executing and advance to the next runnable. It is similar to the `CountDownLatch` mechanism. 

This is not being used on per-test basis but rather on every request that is expected to go through the command dispatcher i.e any request that is going to be run on a different thread. So we use flushScheduler to advance in that case otherwise Robolectric tests will just be hung forever.

So essentially it can't go on the `@After` block as it would have to be done on the level of each token request or other API execution.

Reply:
Thanks for clarifying, @shahzaibj -- resolving this